File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.cc
#include<iostream>
using namespace std;

int operands[11];
int operators[4];
int n;

int MAX=-1000000001;
int MIN = 1000000001;
void solve(int calc, int index) {
	if (index == n) {
		if (calc > MAX) {
			MAX = calc;
		}
		if (calc < MIN) {
			MIN = calc;
		}
		return;
	}

	for (int i = 0; i < 4; i++) {
		if (operators[i] >= 1) {
			operators[i]--;
			if (i == 0) {
				solve(calc + operands[index], index+1);
			}
			else if (i == 1) {
				solve(calc - operands[index], index+1);
			}
			else if (i == 2) {
				solve(calc * operands[index], index+1);
			}
			else if (i == 3) {
				solve(calc / operands[index], index+1);
			}
			operators[i]++;
		}

	}




}


int main() {
	
	cin >> n;
	
	for (int i = 0; i < n; i++) {
		cin >> operands[i];
	}
	for (int i = 0; i < 4; i++) {
		cin >> operators[i];
	}
	solve(operands[0], 1);
	cout << MAX << endl << MIN << endl;


	return 0;
}

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.util.*;
public class Main {
    static int[]num;
    static int N;
    static int[]oper;
    static int max_result=Integer.MIN_VALUE;
    static int min_result=Integer.MAX_VALUE;
    static int result;
    public static void main(String[] args){
        Scanner sc=new Scanner(System.in);
        N=sc.nextInt();
        num=new int[N];
        for(int i=0;i<N;i++){
            num[i]=sc.nextInt();
        }
        oper=new int[4];
        for(int i=0;i<4;i++){
            oper[i]=sc.nextInt();
        }

        dfs(num[0],1);
        System.out.println(max_result+" "+min_result);

    }
    static void dfs(int result, int index){
        if(index==N){
            max_result=Math.max(max_result,result);
            min_result=Math.min(min_result,result);
            return;
        }
        for(int i=0;i<4;i++){
            if(oper[i]>0) {
                oper[i]--;
                if (i == 0) {
                    dfs(result + num[index], index + 1);
                    oper[i]++;

                } else if (i == 1) {
                    dfs(result - num[index], index + 1);
                    oper[i]++;

                } else if (i == 2) {
                    dfs(result * num[index], index + 1);
                    oper[i]++;

                } else if (i == 3) {
                    dfs(result / num[index], index + 1);
                    oper[i]++;

                }
            }

        }

    }
}